# 2. Cleaning

## Set up

In [39]:
from pathlib import Path
import sys

def find_project_root(marker="README.md"):
    current = Path.cwd().resolve()
    for parent in [current, *current.parents]:
        if (parent / marker).exists():
            return parent
    raise FileNotFoundError(f"Could not find project root with marker: {marker}")

PROJECT_ROOT = find_project_root()

if str(PROJECT_ROOT) not in sys.path:
    sys.path.append(str(PROJECT_ROOT))

print(f"Project root: {PROJECT_ROOT}")

Project root: /Users/Local/REPOSITORIES/injury_risk_classifier


In [40]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt

%load_ext autoreload
%autoreload 2

from src.cleaning_and_features import clean

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [41]:
data_path = PROJECT_ROOT / "data" / "interim" / "raw_modeling_data.csv"

df = pd.read_csv(data_path)

In [42]:
df.columns

Index(['top_traffic_accident_offense', 'first_occurrence_date',
       'incident_address', 'geo_lon', 'geo_lat', 'TU1_TRAVEL_DIRECTION',
       'TU2_TRAVEL_DIRECTION', 'TU1_DRIVER_ACTION', 'TU2_DRIVER_ACTION',
       'TU1_VEHICLE_TYPE', 'TU2_VEHICLE_TYPE', 'TU1_DRIVER_HUMANCONTRIBFACTOR',
       'TU2_DRIVER_HUMANCONTRIBFACTOR', 'TU1_PEDESTRIAN_ACTION',
       'SERIOUSLY_INJURED', 'FATALITIES', 'FATALITY_MODE_1', 'FATALITY_MODE_2',
       'SERIOUSLY_INJURED_MODE_1', 'SERIOUSLY_INJURED_MODE_2',
       'TU2_PEDESTRIAN_ACTION', 'ROAD_DESCRIPTION', 'ROAD_CONDITION',
       'LIGHT_CONDITION'],
      dtype='str')

## Cleaning

Code for this is in `clean.py`, but I walk through with the dataset 

In [43]:
__, miss_report = clean.explore(df, missing = True, threshold= 75)

inspecting data...

There are 282358 rows and 24 columns
There 68 duplicate rows



In [44]:
df = df.drop_duplicates()

In [45]:
print(miss_report)

                              data_types  missing_count  missing_percent  \
top_traffic_accident_offense         str              0             0.00   
first_occurrence_date            float64              0             0.00   
incident_address                     str              0             0.00   
geo_lon                          float64          11155             3.95   
geo_lat                          float64          11155             3.95   
TU1_TRAVEL_DIRECTION                 str          10993             3.89   
TU2_TRAVEL_DIRECTION                 str          12629             4.47   
TU1_DRIVER_ACTION                    str          13423             4.75   
TU2_DRIVER_ACTION                    str          15146             5.36   
TU1_VEHICLE_TYPE                     str           6849             2.43   
TU2_VEHICLE_TYPE                     str          10038             3.56   
TU1_DRIVER_HUMANCONTRIBFACTOR        str          15813             5.60   
TU2_DRIVER_H

### Date columns

Turns the julian `first_occurance_date` into two columns, time and date

In [46]:
df = clean.split_datetime(df, "first_occurrence_date")

In [72]:
df["road_description"].value_counts()

road_description
at intersection             88836
non intersection            82112
non-intersection            33467
intersection related        24863
parking lot                 21385
highway interchange         10602
driveway access related      6828
                             5364
alley related                3054
under investigation          1617
ramp related                 1106
ramp                         1038
roundabout                    816
express/managed/hov lane      265
mid-block crosswalk           157
crossover-related              97
auxiliary lane                 78
share-use path or trail        30
Name: count, dtype: int64[pyarrow]

### Text columns

In [48]:
col = ['TU1_TRAVEL_DIRECTION', 'TU2_TRAVEL_DIRECTION', 'TU1_VEHICLE_TYPE', 'TU2_VEHICLE_TYPE', 'TU1_DRIVER_ACTION', 'TU2_DRIVER_ACTION', 
       'TU1_DRIVER_HUMANCONTRIBFACTOR', 'TU2_DRIVER_HUMANCONTRIBFACTOR', 
       'TU1_PEDESTRIAN_ACTION', "ROAD_DESCRIPTION", "ROAD_CONDITION", "LIGHT_CONDITION"]

df = clean.clean_text_columns(df, col)

### Column names

In [49]:
df = clean.clean_column_names(df)

### Create Outcome

In [50]:
df = clean.create_outcome(df)

### Dropping columns

In [51]:
sub = df.loc[:, ["fatality_mode_1", "fatality_mode_2",
            "seriously_injured_mode_1", "seriously_injured_mode_2"]]

In [52]:
__, miss =clean.explore(sub, missing=True)

inspecting data...

There are 282290 rows and 4 columns
There 282084 duplicate rows



There is little unique information gained from these columns. Additionally, I drop these to avoid leakage as the information may not be known till after the outcome was reported.

In [53]:
print(miss)

                         data_types  missing_count  missing_percent  \
fatality_mode_1                 str          78500            27.81   
fatality_mode_2                 str          78661            27.87   
seriously_injured_mode_1        str          77069            27.30   
seriously_injured_mode_2        str          78483            27.80   

                          high missing  
fatality_mode_1                  False  
fatality_mode_2                  False  
seriously_injured_mode_1         False  
seriously_injured_mode_2         False  


In [54]:
df.drop(columns = ["fatality_mode_1", "fatality_mode_2",
                   "seriously_injured_mode_1", "seriously_injured_mode_2"], inplace = True)

In [55]:
df.head(20)

,top_traffic_accident_offense,first_occurrence_date,incident_address,geo_lon,geo_lat,tu1_travel_direction,tu2_travel_direction,tu1_driver_action,tu2_driver_action,tu1_vehicle_type,...,seriously_injured,fatalities,tu2_pedestrian_action,road_description,road_condition,light_condition,datetime,date,time,injured
0,TRAF - ACCIDENT,2.459922e+06,I25 HWYNB / W 6TH AVE,-105.013229,39.725677,north,north,no contributing action,no contributing action,pickup truck/utility van,...,0.0,0.0,NaN,non-intersection,dry,daylight,2022-12-08 14:05:59.999991956,2022-12-08,14:05:59,0
1,TRAF - ACCIDENT,2.459925e+06,E 17TH AVE / N PENNSYLVANIA ST,-104.981084,39.743271,south,east,failed to yield row,no contributing action,passenger car/passenger van,...,0.0,0.0,NaN,at intersection,dry,daylight,2022-12-11 11:39:59.999991056,2022-12-11,11:39:59,0
2,TRAF - ACCIDENT,2.459925e+06,I25 HWYSB / PARK AVEW,-104.994972,39.768330,south,south,followed too closely,no contributing action,passenger car/passenger van,...,0.0,0.0,NaN,non-intersection,dry,daylight,2022-12-11 12:10:00.000004471,2022-12-11,12:10:00,0
3,TRAF - ACCIDENT - POLICE,2.459922e+06,I25 HWYNB / 20TH ST,-105.004334,39.760949,south,south,followed too closely,no contributing action,pickup truck/utility van,...,0.0,0.0,NaN,non-intersection,dry,dark-lighted,2022-12-08 18:26:59.999983903,2022-12-08,18:26:59,0
4,TRAF - ACCIDENT - HIT & RUN,2.459922e+06,W 37TH AVE / N PECOS ST,-105.006450,39.768054,<NA>,<NA>,<NA>,<NA>,unknown (hit and run only),...,0.0,0.0,NaN,non-intersection,dry,dark-lighted,2022-12-08 19:30:00.000000000,2022-12-08,19:30:00,0
5,TRAF - ACCIDENT - HIT & RUN,2.459916e+06,S OSCEOLA ST / W BAYAUD AVE,-105.038499,39.714959,south,north,careless driving,no contributing action,passenger car/passenger van,...,0.0,0.0,NaN,non-intersection,dry,dark-lighted,2022-12-02 02:00:00.000013415,2022-12-02,02:00:00,0
6,TRAF - ACCIDENT - HIT & RUN,2.459916e+06,300 BLOCK E 17TH AVE,-104.983103,39.743008,<NA>,northeast,<NA>,no contributing action,<NA>,...,0.0,0.0,NaN,non-intersection,dry,daylight,2022-12-02 12:20:00.000008943,2022-12-02,12:20:00,0
7,TRAF - ACCIDENT - HIT & RUN,2.459916e+06,N OGDEN ST / E 11TH AVE,-104.975278,39.733689,south,east,disregarded stop sign,no contributing action,pickup truck/utility van,...,0.0,0.0,NaN,at intersection,dry,dark-lighted,2022-12-02 15:59:59.999986584,2022-12-02,15:59:59,0
8,TRAF - ACCIDENT,2.459916e+06,N CENTRAL PARK BLVD / N VERBENA ST,-104.889926,39.751852,south,south,careless driving,lane violation,passenger car/passenger van,...,0.0,0.0,NaN,intersection related,dry,daylight,2022-12-02 15:45:00.000000000,2022-12-02,15:45:00,0
9,TRAF - ACCIDENT - HIT & RUN,2.459916e+06,I70 HWYWB / N NORTHFIELD QUEBEC ST,-104.903436,39.778421,north,north,lane violation,no contributing action,passenger car/passenger van,...,0.0,0.0,NaN,intersection related,dry,dark-lighted,2022-12-02 19:29:00.000003572,2022-12-02,19:29:00,0


## Recoding

The following cells recode the messy text columns to understand the columns and combine classes. For example, distracted coveres "distracted-other", "distracted/other interior", "distracted/other exterior". 

This is completed for the following sets of columns 

| Information | Column names | Descrip. |  Example|
| -------- | -------- | -------- | -------- |
| vehicle_type | `tu1_vehicle_type`, `tu2_vehicle_type` | Vehicle Type |  "passenger", "vehicle over 10000 lbs" |
| driver actions | `tu1_driver_action`, `tu2_driver_action` | Driver actions, specifically in vehicle operation | "failure to yeild" |
| human factor | `tu1_driver_humancontribfactor`, `tu2_driver_humancontribfactor` | Driver actions, not directly in vehicle movement | "distracted", "medical" |

For each pair of columns, I 
1. count how often each entry appears per column (`create_types_table`)
2. combine related entries (`bin_`) 
3. try to understand how many of each accident type, based on binned columns, were injurious. 

### Vehicle Types 

In [56]:
vehicle_types = clean.create_types_tables(
    df,
    "tu1_vehicle_type",
    "tu2_vehicle_type",
    first_label="first_vehicle_type",
    second_label="second_vehicle_type"
)

vehicle_types

,first_vehicle_type,second_vehicle_type,total
passenger car/van,96819,97347,194166
suv,59399,64389,123788
passenger car/passenger van,34815,37402,72217
pickup truck/utility van,31487,28162,59649
other,8104,16253,24357
<NA>,6849,10037,16886
hit and run unknown,14703,483,15186
,2130,8946,11076
vehicle over 10000 lbs,6324,4630,10954
pickup truck/utility van with trailier,4680,3866,8546


In [57]:
df = clean.text_binning(df = df,
        columns = clean.vehicle_type_cols,
        mapping= clean.VEHICLE_TYPE_MAP,
        default_value="other",
        unknown_value="unknown")

In [58]:
clean.accident_summary_wrt(df, col= "tu1_vehicle_type_binned")

,count,mean
tu1_vehicle_type_binned,,
motorcycle,1925,0.357922
low-speed,59,0.220339
bicycle,155,0.141935
other,8462,0.058142
passenger,131634,0.021917
suv,59577,0.019722
pickup truck,31487,0.018293
unknown,8979,0.015146
heavy vehicle,8733,0.011794


Out of the 1925 accidents where the first vehicle was a motorcycle, around 35% of those were associated with a severe injury or fatality. 

### Driver actions

In [59]:
driver_action = clean.create_types_tables(
    df,
    "tu1_driver_action",
    "tu2_driver_action",
    first_label="first_driver_action",
    second_label="first_driver_action"
)

driver_action

,first_driver_action,total
tu2_driver_action,,
other,91525,183050
no action,63197,126394
no contributing action,61370,122740
00,29025,58050
<NA>,15145,30290
,14136,28272
under investigation,1617,3234
lane violation,1106,2212
followed too closely,969,1938


In [60]:
df = clean.text_binning(df = df,
        columns = clean.driver_action_cols,
        mapping= clean.DRIVER_ACTION_MAP,
        default_value="other",
        unknown_value="unknown")

In [61]:
clean.accident_summary_wrt(df, col= "tu1_driver_action_binned")

,count,mean
tu1_driver_action_binned,,
reckless_or_speed_related,15527,0.040574
failure_to_yield_or_obey_control,42916,0.031643
other_or_invalid,36692,0.025619
careless_or_loss_of_control,102842,0.025525
traffic_flow_or_position,8761,0.021345
turning,10046,0.015927
no_action,11428,0.015926
unknown,19211,0.013117
lane_or_passing,23679,0.008108


### Human contributing factor

In [62]:
human_factor = clean.create_types_tables(
    df,
    "tu1_driver_humancontribfactor",
    "tu2_driver_humancontribfactor",
    first_label="first_driver",
    second_label="second_driver"
)

human_factor

,first_driver,second_driver,total
no apparent,78474,158520,236994
no apparent contributing factor,16373,60316,76689
other,37660,26067,63727
<NA>,15813,14181,29994
aggressive driving,23842,912,24754
,8808,15310,24118
distracted-other,16721,324,17045
driver inexperience,15977,525,16502
not observed,12798,2813,15611
dui/dwai/duid,10711,186,10897


In [63]:
df = clean.text_binning(df = df,
        columns = clean.human_contrib_cols,
        mapping= clean.HUMAN_CONTRIB_MAP,
        default_value="other_or_invalid",
        unknown_value="other_or_invalid")

In [64]:
clean.accident_summary_wrt(df, col= "tu1_driver_humancontribfactor_binned")

,count,mean
tu1_driver_humancontribfactor_binned,,
chase,798,0.073935
impairment,10711,0.056484
medical_or_fatigue,5965,0.046102
aggressive,23842,0.042153
unknown_or_not_observed,41830,0.026584
inexperience_or_unfamiliar,26852,0.022419
did_not_see,8490,0.021555
distracted,29561,0.019384
no apparent,107645,0.016787


Among the 5965 crashes where driver 1 had a medical/ability-related contributing factor, about 4.6% resulted in injury/fatality.

However, this does not mean the medical factors caused the accident or caused the accident top be injurious. Of the crashed that could be categorized, 4.6% involved injury. 

In [67]:
df["tu1_travel_direction"].unique()


<ArrowStringArray>
[              'north',                'east',               'south',
                  <NA>,           'northeast', 'under investigation',
                'west',           'southwest',           'northwest',
           'southeast',               'other',                    '',
                  '01',                  '05',                  '03',
                  '07',                  '02',                  '00']
Length: 18, dtype: string

In [68]:
df["tu2_travel_direction"].unique()

<ArrowStringArray>
[              'north',                'east',               'south',
                  <NA>,           'northeast', 'under investigation',
                'west',           'southwest',           'northwest',
           'southeast',               'other',                    '',
                  '01',                  '05',                  '03',
                  '07',                  '02',                  '00']
Length: 18, dtype: string

# parquet

In [ ]:
import os 
os.chdir('..') 

In [ ]:
pq = pd.read_parquet("data/interim/cleaner_data.parquet")

In [ ]:
pq.columns

Index(['top_traffic_accident_offense', 'first_occurrence_date',
       'incident_address', 'lon', 'lat', 'tu1_travel_direction',
       'tu2_travel_direction', 'road_description', 'road_condition',
       'light_condition', 'datetime', 'date', 'time', 'injured',
       'tu1_vehicle_type_binned', 'tu2_vehicle_type_binned',
       'tu1_driver_humancontribfactor_binned',
       'tu2_driver_humancontribfactor_binned', 'tu1_driver_action_binned',
       'tu2_driver_action_binned'],
      dtype='str')